packages/adapter/src/io-ts.ts

```diff
@@ -0,0 +1,62 @@
+import * as t from "io-ts/lib/index.js";
+
+/**
+ * In the TS-SDK we duplicate the type and guard definition for each type as the
+ * inferred type from io-ts does not produce a good type export when used with
+ * typedoc. Sadly this means we have to maintain two definitions for each type which
+ * can lead to errors, so we need a mechanism to assert that the type and guard are
+ * equal.
+ * Normally we could simply do:
+ ```
+ interface Foo {
+    prop: string;
+ }
+ const FooGuard: t.Type<Foo> = t.type({
+    prop: t.string
+ })
+ ```
+ * And the explicit type assertion in the FooGuard ensure us that the type are the same.
+ * But when we include Branded or Newtype types in the definition, the full type is not the same
+ ```
+ interface Foo {
+    prop: ContractId;
+ }
+ // This throws a type error
+ const FooGuard: t.Type<Foo> = t.type({
+    prop: ContractIdGuard
+ })
+ ```
+ *
+ * To understand the error we need to know that  `t.Type<A> = t.Type<A, O = A, I = unknown>`
+ * where `A` is the underlying type, `O` is the Output type that we get when
+ * we call `FooGuard.encode` and `I` is the Input type we expect when we call `FooGuard.decode`.
+ * When we use a Branded or Newtype guard, the Output type is different from the underlying type,
+ * so we get an error.
+ *
+ * The `assertGuardEqual` function can be used to only assert on the representation type A and leave
+ * the Input and Output type unchanged.
+ *
+ ```
+ interface Foo {
+    prop: ContractId;
+ }
+ const FooGuard = assertGuardEqual(proxy<Foo>(), t.type({
+    prop: ContractIdGuard
+ }))
+ ```
+ */
+
+export function assertGuardEqual<A, G extends t.Type<A, any, any>>(
+  proxy: A,
+  guard: G
+): G {
+  return guard;
+}
+
+/**
+ * Utility function to set up an unused value with an explicit type A.
+ * @returns
+ */
+export function proxy<A = never>(): A {
+  return null as any;
+}
```

Comment:
After making this I realized that I should probably try to make a `Proxy<T>` type that is unusable instead of casting to A. For the purpose that is being used it shouldn't matter, but it would be better of course

Reply:
I'm assuming this is missing functionality in `fp-ts` and `io-ts` and that it corresponds to [`data Proxy (t :: k)`](https://hackage.haskell.org/package/base-4.19.0.0/docs/Data-Proxy.html#t:Proxy) from Haskell.

Can we live without this? It looks like growing tech debt. Is there an alternate solution where we cut out some fp-ts code instead which makes this unecessary?

Reply:
It is not missing functionality from fp-ts/io-ts in the sense that this is not required for "normal use". It is required by us for the reason I wrote at the beginning of the file.

If you use TypeScript (or any statically typed language) you'll eventually need some sort of dynamic checks or you are bound to have runtime errors. 

For TS as far as I know you have:
* io-ts
* [runtypes](https://github.com/pelotom/runtypes)
* [zod](https://github.com/colinhacks/zod)

and possible others...

Back in the day I created [parmenides](https://github.com/dggluz/parmenides) with a friend before the other libraries existed/got traction.

Knowing that we need runtime validation, and we already have io-ts in place (which has all the features we need and more), I don't believe this is adding to the technical debt. We just need to be judicious on what and how we expose things. See this [exported doc](https://input-output-hk.github.io/marlowe-ts-sdk/modules/_marlowe_io_language_core_v1.html) as an example and note the difference between the types and the dynamic guards.